### PR TITLE
[CARBONDATA-3800] Load data to SI and MV after insert stage command

### PR DESCRIPTION
**Pull request by Indhumathi27**
### Why is this PR needed?
1. Data is not loaded to child tables(SI and MV) after executing insert stage command.
2. Datamap keyword still exists in some files
 
 ### What changes were proposed in this PR?
1. Add Load Pre and Post listener's in CarbonInsertStageCommand to trigger data load to Secondary indexes and materialized views.
2. Rename datamap to index
    
 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - Yes

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2949/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1231/

File: integration/spark/src/main/scala/org/apache/spark/sql/events/MergeIndexEventListener.scala [hunk 1]
```diff
@@ -43,6 +43,11 @@ class MergeIndexEventListener extends OperationEventListener with Logging {
   override def onEvent(event: Event, operationContext: OperationContext): Unit = {
     event match {
       case preStatusUpdateEvent: LoadTablePreStatusUpdateEvent =>
+        // skip merge index in case of insert stage flow
+        if (null != operationContext.getProperty("isInsertStage") &&
```
Comment (niuge01) [inline integration/spark/src/main/scala/org/apache/spark/sql/events/MergeIndexEventListener.scala:47]: Define "isInsertStage" as a constant will be better.

**Comment by niuge01** [comment]
so much control properties

File: integration/spark/src/main/scala/org/apache/spark/sql/events/MergeIndexEventListener.scala (hunk 1 above)
Comment (Indhumathi27) [reply integration/spark/src/main/scala/org/apache/spark/sql/events/MergeIndexEventListener.scala]: done

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2957/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2962/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1243/

**Comment by Indhumathi27** [comment]
@akashrn5 @kunal642 Please help to review and merge

**Comment by akashrn5** [comment]
@Indhumathi27 please resolve conflicts

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1273/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2992/

**Comment by Indhumathi27** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2999/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1280/

File: integration/flink/src/test/scala/org/apache/carbon/flink/TestCarbonPartitionWriter.scala [hunk 2]
```diff
@@ -206,6 +210,184 @@ class TestCarbonPartitionWriter extends QueryTest {
...
+      var isFilterHitSecondaryIndex = false
+      df.queryExecution.sparkPlan.transform {
+        case broadCastSIFilterPushDown: BroadCastSIFilterPushJoin =>
+          isFilterHitSecondaryIndex = true
+          broadCastSIFilterPushDown
+      }
+      assert(isFilterHitSecondaryIndex)
+
+    } finally {
+      CarbonProperties.getInstance().removeProperty(CarbonCommonConstants.ENABLE_QUERY_STATISTICS)
```
Comment (kunal642) [inline integration/flink/src/test/scala/org/apache/carbon/flink/TestCarbonPartitionWriter.scala:300]: Why is this done here and not in afterAll()?

Comment (Indhumathi27) [reply integration/flink/src/test/scala/org/apache/carbon/flink/TestCarbonPartitionWriter.scala]: refactored. Please check

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3005/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1286/

**Comment by QiangCai** [comment]
I search "DataMap" with case-insensitive and ignore the format module. 
still found the following 44 usages.

```
Targets
    Occurrences of 'DataMap' in project
Found Occurrences  (44 usages found)
    Production  (22 usages found)
        carbondata-core  (1 usage found)
            org.apache.carbondata.core.indexstore  (1 usage found)
                BlockletIndexStore.java  (1 usage found)
                    put(TableBlockIndexUniqueIdentifierWrapper, BlockletIndexWrapper)  (1 usage found)
                        251 // As dataMap will use unsafe memory, it is not recommended to overwrite an existing entry
        carbondata-parent  (3 usages found)
            docs  (1 usage found)
                configuration-parameters.md  (1 usage found)
                    44 datamap
            integration/spark-common-cluster-test/src/test/scala/org/apache/carbondata/cluster/sdv/generated  (2 usages found)
                LuceneTestCase.scala  (2 usages found)
                    34 sql("DROP TABLE IF EXISTS datamap_main")
                    38 test("LuceneDataMap_TC001", Include) {
        carbondata-spark_2.3  (18 usages found)
            org.apache.spark.sql  (1 usage found)
                CarbonSession.scala  (1 usage found)
                    106 // explain command will output the dataMap information only if enable.query.statistics = true
            org.apache.spark.sql.events  (2 usages found)
                MergeIndexEventListener.scala  (2 usages found)
                    100 // clear Block dataMap Cache
                    158 // clear Block dataMap Cache
            org.apache.spark.sql.execution.command.cache  (8 usages found)
                CarbonShowCacheCommand.scala  (8 usages found)
                    71 AttributeReference("Datamap size", StringType, nullable = false)(),
                    396 // datamapName -> (datamapProviderName, indexSize, datamapSize)
                    407 var (allIndexSize, allDatamapSize) = (0L, 0L)
                    418 allDatamapSize += cacheable.getMemorySize
                    421 (allIndexSize, allDatamapSize)
                    452 var (allIndexSize, allDatamapSize) = (0L, 0L)
                    458 allDatamapSize += sum
                    463 (allIndexSize, allDatamapSize)
            org.apache.spark.sql.secondaryindex.Jobs  (6 usages found)
                BlockletIndexDetailsWithSchema.java  (3 usages found)
                    BlockletIndexDetailsWithSchema(BlockletIndexWrapper, boolean)  (2 usages found)
                        47 // In one task all dataMaps will have the same cardinality and schema therefore
                        48 // segmentPropertyIndex can be fetched from one dataMap
                    29 * class that holds dataMaps, column cardinality, columnSchema and other related information for
                BlockletIndexInputFormat.java  (1 usage found)
                    createRecordReader(InputSplit, TaskAttemptContext)  (1 usage found)
                        182 // as segmentId will be same for all the dataMaps and segmentProperties cache is
                SparkBlockletIndexLoaderJob.scala  (2 usages found)
                    122 // add dataMap to cache
                    137 * This RDD is used to load the dataMaps of a segment
            org.apache.spark.util  (1 usage found)
                AlterTableUtil.scala  (1 usage found)
                    803 // clear dataMap cache
    Test  (22 usages found)
        carbondata-core  (4 usages found)
            org.apache.carbondata.core.metadata.schema.table  (4 usages found)
                TableInfoTest.java  (4 usages found)
                    testTableInfoDeserializationWithSchemaEvolution()  (1 usage found)
                        130 + "\"isTransactionalTable\":true,\"dataMapSchemaList\":[],"
                    testIndexSchemaDeserializationWithClassName()  (3 usages found)
                        167 + "\"/opt/store/carbonversion_1_3/testinttype3\",\"dataMapSchemaList\":"
                        168 + "[{\"dataMapName\":\"dm1\",\"className\":"
                        169 + "\"org.apache.carbondata.core.datamap.AggregateDataMap\","
        carbondata-examples  (1 usage found)
            org.apache.carbondata.examplesCI  (1 usage found)
                RunExamples.scala  (1 usage found)
                    107 test("LuceneDataMapExample") {
        carbondata-spark_2.3  (17 usages found)
            org.apache.carbondata.index.bloom  (3 usages found)
                BloomCoarseGrainIndexFunctionSuite.scala  (1 usage found)
                    40 val bigFile = s"$resourcesPath/bloom_datamap_function_test_big.csv"
                BloomCoarseGrainIndexSuite.scala  (2 usages found)
                    37 val bigFile = s"$resourcesPath/bloom_datamap_input_big.csv"
                    38 val smallFile = s"$resourcesPath/bloom_datamap_input_small.csv"
            org.apache.carbondata.index.lucene  (14 usages found)
                LuceneCoarseGrainIndexSuite.scala  (7 usages found)
                    29 val file2 = resourcesPath + "/datamap_input.csv"
                    46 sql("DROP TABLE IF EXISTS datamap_test")
                    49         | CREATE TABLE datamap_test(id INT, name STRING, city STRING, age INT)
                    57          | ON datamap_test (name, city)
                    61 sql(s"LOAD DATA LOCAL INPATH '$file2' INTO TABLE datamap_test OPTIONS('header'='false')")
                    63 checkAnswer(sql("select * from datamap_test where name='n502670'"),
                    70 sql("DROP TABLE IF EXISTS datamap_test")
                LuceneFineGrainIndexSuite.scala  (7 usages found)
                    42 val file2 = resourcesPath + "/datamap_input.csv"
                    789 sql("DROP TABLE IF EXISTS datamap_copy")
                    799         | CREATE TABLE datamap_copy(id INT, name STRING, city STRING, age INT)
                    807 sql("insert into datamap_copy select * from index_test4")
                    815 sql(s"select * from datamap_copy where name like '%n%'"))
                    817 sql("drop table datamap_copy")
                    884 sql("DROP TABLE IF EXISTS datamap_main")

```

File: integration/flink/src/test/scala/org/apache/carbon/flink/TestCarbonPartitionWriter.scala [hunk 3]
```diff
@@ -201,14 +126,145 @@ class TestCarbonPartitionWriter extends QueryTest {
...
+      val source = getTestSource(dataCount)
+      executeStreamingEnvironment(tablePath, writerProperties, carbonProperties, environment, source)
+
+      sql(s"INSERT INTO $tableName STAGE")
+
+      checkAnswer(sql("select count(*) from si_1"), Seq(Row(10))  )
+      checkAnswer(sql(s"SELECT count(*) FROM $tableName"), Seq(Row(10)))
+      // check if query hits si
+      val df = sql(s"select count(intField) from $tableName where stringField1 = 'si1'")
+      checkAnswer(df, Seq(Row(1)))
```
Comment (QiangCai) [inline integration/flink/src/test/scala/org/apache/carbon/flink/TestCarbonPartitionWriter.scala:157]: can you check the detailed content too?

File: integration/flink/src/test/scala/org/apache/carbon/flink/TestCarbonWriter.scala [hunk 4]
```diff
@@ -19,85 +19,42 @@ package org.apache.carbon.flink
...
 import org.apache.spark.sql.execution.exchange.Exchange
+import org.apache.spark.sql.secondaryindex.joins.BroadCastSIFilterPushJoin
+import org.scalatest.BeforeAndAfterAll
 
-class TestCarbonWriter extends QueryTest {
+class TestCarbonWriter extends QueryTest with BeforeAndAfterAll{
 
   val tableName = "test_flink"
   val bucketTableName = "insert_bucket_table"
+  val rootPath: String = System.getProperty("user.dir") + "/target/test-classes"
```
Comment (QiangCai) [inline integration/flink/src/test/scala/org/apache/carbon/flink/TestCarbonWriter.scala:45]: How about to use variable 'target' in QueryTest ?

File: integration/flink/src/test/scala/org/apache/carbon/flink/TestCarbonWriter.scala [hunk 5]
```diff
@@ -301,16 +156,156 @@ class TestCarbonWriter extends QueryTest {
...
+      checkAnswer(sql(s"SELECT count(*) FROM $tableName"), Seq(Row(0)))
+      checkAnswer(sql("select count(*) from si_1"), Seq(Row(0)))
+
+      sql(s"INSERT INTO $tableName STAGE")
+
+      checkAnswer(sql("select count(*) from si_1"), Seq(Row(1000)))
+      checkAnswer(sql(s"SELECT count(*) FROM $tableName"), Seq(Row(1000)))
+      // check if query hits si
+      val df = sql(s"select count(intField) from $tableName where stringField1 = 'si12'")
+      checkAnswer(df, Seq(Row(1)))
```
Comment (QiangCai) [inline integration/flink/src/test/scala/org/apache/carbon/flink/TestCarbonWriter.scala:188]: better to check the data also

**Comment by Indhumathi27** [comment]
@QiangCai Fixed all comments

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1293/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3011/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3014/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1296/

**Comment by QiangCai** [comment]
LGTM

**Comment by akashrn5** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1304/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3022/

**Comment by akashrn5** [comment]
LGTM

**Pull request closed without merging**